Navigation within iframe doesn't exit fullscreen for parent iframe element

nt1m opened the pull request:
#### 9215d7d582cb6d40cb439a45cb2157ab9bcbb575
<pre>
Navigation within iframe doesn&apos;t exit fullscreen for parent iframe element
<a href="https://bugs.webkit.org/show_bug.cgi?id=251896">https://bugs.webkit.org/show_bug.cgi?id=251896</a>
rdar://104393093

Reviewed by Chris Dumez.

This was because the exiting document was already detached from the Frame object, causing it to fail to find the top document.
To fix this, we store the top document when initializing FullscreenManager.

* LayoutTests/fullscreen/fullscreen-iframe-navigation-expected.html: Added.
* LayoutTests/fullscreen/fullscreen-iframe-navigation.html: Added.
* LayoutTests/fullscreen/resources/fullscreen-iframe-navigation-target.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::cancelFullscreen):
(WebCore::documentsToUnfullscreen):
(WebCore::FullscreenManager::exitFullscreen):
(WebCore::FullscreenManager::finishExitFullscreen):
(WebCore::FullscreenManager::didExitFullscreen):
* Source/WebCore/dom/FullscreenManager.h:

Canonical link: <a href="https://commits.webkit.org/260024@main">https://commits.webkit.org/260024@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2eff2c98c3f57e90eb412765d411802845f9d9d1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106701 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15692 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39489 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115888 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115482 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110607 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17204 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6929 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98896 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112470 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13067 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96031 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40632 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94939 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27683 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8914 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29034 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9474 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6068 "Too many flaky failures: css3/masking/mask-svg-no-fragmentId-tiled.html, css3/masking/mask-svg-script-none-to-png.html, fast/css/large-number-round-trip.html, fast/css/last-child-display-change.html, gamepad/gamepad-vibration-document-no-longer-fully-active.html, http/tests/appcache/cyrillic-uri.html, http/tests/cache-storage/cache-persistency.https.html, imported/w3c/web-platform-tests/css/css-pseudo/first-line-and-marker.html, imported/w3c/web-platform-tests/css/css-pseudo/first-line-line-height-001.html, imported/w3c/web-platform-tests/css/selectors/invalidation/media-loading-pseudo-classes-in-has.html, imported/w3c/web-platform-tests/css/selectors/invalidation/media-pseudo-classes-in-has.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15087 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48583 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6924 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10996 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -51,7 +51,8 @@ namespace WebCore {
 FullscreenManager::FullscreenManager(Document& document)
-    : m_document { document }
+    : m_document(document)
+    , m_topDocument(&document.topDocument())
```
No ampersand. No point in WeakPtr null checking a pointer for no reason.

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -256,8 +257,8 @@ void FullscreenManager::cancelFullscreen()
     // context's document and subsequently empty that document's fullscreen element stack."
-    Document& topDocument = document().topDocument();
-    if (!topDocument.fullscreenManager().fullscreenElement()) {
+    Ref topDocument = m_topDocument ? *m_topDocument.get() : document();
```
`.get()` shouldn't be needed.

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -274,24 +275,24 @@ void FullscreenManager::cancelFullscreen()
 
 // https://fullscreen.spec.whatwg.org/#collect-documents-to-unfullscreen
-static Vector<Ref<Document>> documentsToUnfullscreen(Document& firstDocument)
+static Vector<Ref<Document>> documentsToUnfullscreen(const Ref<Document>& firstDocument)
```
cc @rniwa on this to see if this is the new preferred style. I liked the old code better personally.

cdumez replied:
I heard back for Ryosuke, the old style is still preferred. Please omit this change.

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -274,24 +275,24 @@ void FullscreenManager::cancelFullscreen()
+static Vector<Ref<Document>> documentsToUnfullscreen(const Ref<Document>& firstDocument)
 {
-    Vector<Ref<Document>> documents { Ref { firstDocument } };
+    Vector<Ref<Document>> documents { Ref { firstDocument.get() } };
```
`Vector<Ref<Document>> documents { firstDocument.copyRef() };`

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -312,17 +313,17 @@ void FullscreenManager::exitFullscreen(RefPtr<DeferredPromise>&& promise)
     INFO_LOG(LOGIDENTIFIER);
 
-    auto* exitingDocument = &document();
+    RefPtr exitingDocument = &document();
```
Should use `Ref` since it can't be null

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -312,17 +313,17 @@ void FullscreenManager::exitFullscreen(RefPtr<DeferredPromise>&& promise)
     auto mode = ExitMode::NoResize;
     auto exitDocuments = documentsToUnfullscreen(*exitingDocument);
-    auto& topDocument = this->topDocument();
+    Ref topDocument = m_topDocument ? *m_topDocument.get() : document();
```
`.get()` is not needed

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -312,17 +313,17 @@ void FullscreenManager::exitFullscreen(RefPtr<DeferredPromise>&& promise)
+    if (exitsTopDocument && topDocument->fullscreenManager().isSimpleFullscreenDocument()) {
         mode = ExitMode::Resize;
-        exitingDocument = &topDocument;
+        exitingDocument = topDocument.ptr();
```
`exitingDocument = topDocument;`

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -369,7 +370,7 @@ void FullscreenManager::exitFullscreen(RefPtr<DeferredPromise>&& promise)
             page->chrome().client().exitFullScreenForElement(m_fullscreenElement.get());
         else {
-            finishExitFullscreen(document(), ExitMode::NoResize);
+            finishExitFullscreen(Ref { document() }, ExitMode::NoResize);
```
Is this really needed?

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -378,14 +379,14 @@ void FullscreenManager::exitFullscreen(RefPtr<DeferredPromise>&& promise)
 }
 
-void FullscreenManager::finishExitFullscreen(Document& currentDocument, ExitMode mode)
+void FullscreenManager::finishExitFullscreen(const Ref<Document>& currentDocument, ExitMode mode)
```
cc @rniwa for his opinion on this.

cdumez replied:
I heard back for Ryosuke, the old style is still preferred. Please omit this change.

cdumez commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -541,7 +542,8 @@ bool FullscreenManager::didExitFullscreen()
     INFO_LOG(LOGIDENTIFIER);
 
-    finishExitFullscreen(topDocument(), ExitMode::Resize);
+    Ref topDocument = m_topDocument ? *m_topDocument.get() : document();
```
`.get()` shouldn't be needed

cdumez commented on Source/WebCore/dom/FullscreenManager.h:
```diff
@@ -47,7 +47,6 @@ class FullscreenManager final : public CanMakeWeakPtr<FullscreenManager> {
 
     Document& document() { return m_document; }
     const Document& document() const { return m_document; }
-    Document& topDocument() const { return m_document.topDocument(); }
```
Given how it is used, we may want to make this getter private instead and have it return `document()` when `m_topDocument` is null. It would simplify the call sites.

cdumez reviewed (approved):
LGTM if the bots are happy

cdumez commented on Source/WebCore/dom/FullscreenManager.h:
```diff
@@ -108,7 +107,10 @@ class FullscreenManager final : public CanMakeWeakPtr<FullscreenManager> {
     WTFLogChannel& logChannel() const;
 #endif
 
+    Document& topDocument() { return m_topDocument ? *m_topDocument : document().topDocument(); }
```
I think this would look clearer if this was:
`Document& topDocument() { return m_topDocument ? *m_topDocument : m_document; }`

It was like this in an earlier iteration I think. Not sure why you made this change (I don't think it should be needed).

nt1m committed:
Navigation within iframe doesn't exit fullscreen for parent iframe element
https://bugs.webkit.org/show_bug.cgi?id=251896
rdar://104393093

Reviewed by Chris Dumez.

This was because the exiting document was already detached from the Frame object, causing it to fail to find the top document.
To fix this, we store the top document when initializing FullscreenManager.

* LayoutTests/fullscreen/fullscreen-iframe-navigation-expected.html: Added.
* LayoutTests/fullscreen/fullscreen-iframe-navigation.html: Added.
* LayoutTests/fullscreen/resources/fullscreen-iframe-navigation-target.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::cancelFullscreen):
(WebCore::documentsToUnfullscreen):
(WebCore::FullscreenManager::exitFullscreen):
(WebCore::FullscreenManager::finishExitFullscreen):
(WebCore::FullscreenManager::didExitFullscreen):
* Source/WebCore/dom/FullscreenManager.h:

Canonical link: https://commits.webkit.org/260024@main
